File: patterns/java/src/refactoring_guru/patterns/iterator/example/FacebookIterator.java
package refactoring_guru.patterns.iterator.example;

public class FacebookIterator implements ProfileIterator {
    private Facebook facebook;
    private int profileId;
    private String type;
    private int currentPosition;
    private Profile[] cache;

    public FacebookIterator(Facebook facebook, int profileId, String type) {
        this.facebook = facebook;
        this.profileId = profileId;
        this.type = type;
    }

    private void initIfNeeded() {
        if (cache == null) {
            cache = facebook.sendSophisticatedSocialGraphRequest(profileId);
        }
    }

    @Override
    public boolean hasNext() {
        initIfNeeded();
        return cache.length < currentPosition;
    }

    @Override
    public Profile getNext() {
        if (hasNext()) {
            currentPosition++;
            return cache[currentPosition];
        }
        return null;
    }
}

File: patterns/cs/src/RefactoringGuru/Iterator/Example/Iterators/FacebookIterator.cs
using System;
using System.Collections.Generic;
using RefactoringGuru.Iterator.Example.SocialNetworks;

namespace RefactoringGuru.Iterator.Example.Iterators
{
    class FacebookIterator : ProfileIterator
    {
        private Facebook facebook;
        private String type;
        private String email;
        private int currentPosition = 0;
        private List<String> emails = new List<String>();
        private List<Profile.Profile> profiles = new List<Profile.Profile>();

        public FacebookIterator(Facebook facebook, String type, String email)
        {
            this.facebook = facebook;
            this.type = type;
            this.email = email;
        }

        private void LazyLoad()
        {
            if (emails.Count == 0)
            {
                List<String> profiles = facebook.RequestProfileFriendsFromFacebook(this.email, this.type);
                foreach (var profile in profiles)
                {
                    this.emails.Add(profile);
                    this.profiles.Add(null);
                }
            }
        }

        public bool HasNext()
        {
            LazyLoad();
            return currentPosition < emails.Count;
        }

        public Profile.Profile GetNext()
        {
            if (!HasNext())
            {
                return null;
            }

            String friendEmail = emails[currentPosition];
            Profile.Profile friendProfile = profiles[currentPosition];
            if (friendProfile == null)
            {
                friendProfile = facebook.RequestProfileFromFacebook(friendEmail);
                profiles.Insert(currentPosition, friendProfile);
            }
            currentPosition++;
            return friendProfile;
        }

        public void Reset()
        {
            currentPosition = 0;
        }
    }
}
